go/database/mpt/proof_test.go

```diff
@@ -753,12 +753,183 @@ func TestWitnessProof_Extract_EmbeddedNode_In_Proof(t *testing.T) {
 			if got, want := complete, true; got != want {
 				t.Errorf("unexpected proof existence: got %v, want %v", got, want)
 			}
-			if got, want := proof, totalProof; !got.Equals(want) {
+			if got, want := proof, totalProof; !want.Equals(got) {
 				t.Errorf("unexpected proof: got %v, want %v", got, want)
 			}
 		})
 	}
+}
+
+func TestCreateWitnessProof_GetStorageElements(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	address := common.Address{1}
+	key := common.Key{2}
+
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+	addressNibbles := AddressToNibblePath(address, ctxt)
+	keyNibbles := KeyToNibblePath(key, ctxt)
+
+	desc := &Branch{
+		children: Children{
+			addressNibbles[0]: &Branch{
+				children: Children{
+					addressNibbles[1]: &Extension{
+						path: addressNibbles[2:50],
+						next: &Tag{"A", &Account{address: address, pathLength: 14, info: AccountInfo{common.Nonce{1}, amount.New(1), common.Hash{0xAA}},
+							storage: &Tag{"B", &Branch{
+								children: Children{
+									keyNibbles[0]: &Extension{path: keyNibbles[1:40], next: &Value{key: key, length: 24, value: common.Value{0x12}}},
+								}}}}}},
+				}}},
+	}
+
+	root, _ := ctxt.Build(desc)
+
+	proof, err := CreateWitnessProof(ctxt, &root, address, key)
+	if err != nil {
+		t.Fatalf("failed to create proof: %v", err)
+	}
+
+	if !proof.IsValid() {
+		t.Fatalf("proof is not valid")
+	}
+
+	accountProofWant, err := CreateWitnessProof(ctxt, &root, address)
+	if err != nil {
+		t.Fatalf("failed to create account proof: %v", err)
+	}
+
+	// intersect complete proof and account only proof -> it gives the storage proof
+	storageProofWant := WitnessProof{make(proofDb)}
+	for k, v := range proof.proofDb {
+		if _, exists := accountProofWant.proofDb[k]; !exists {
+			storageProofWant.proofDb[k] = v
+		}
+	}
+
+	_, accountNode := ctxt.Get("A")
+	accountHandle := accountNode.GetViewHandle()
+	storageHashWant := accountHandle.Get().(*AccountNode).storageHash
+	accountHandle.Release()
+
+	hash, _ := ctxt.getHashFor(&root)
+
+	accountProof, complete := proof.Extract(hash, address)
+	if !complete {
+		t.Fatalf("proof is not complete")
+	}
+	if !accountProof.IsValid() {
+		t.Fatalf("account proof is not valid")
+	}
+
+	t.Run("Extract storage", func(t *testing.T) {
+		storageElements, storageHash, complete := proof.GetStorageElements(hash, address, key)
+		if !complete {
+			t.Fatalf("proof is not complete")
+		}
+		storageProof := CreateWitnessProofFromNodes(storageElements)
+		if !storageProof.IsValid() {
+			t.Fatalf("storage proof is not valid")
+		}
+
```

Comment:
Maybe it should be tested that
- the list of `storageElements` is not empty
- one of the `storageElements` has a hash that is equivalent to the `storageHash`

Reply:
good idea, will add